docs/HowToFAQ/Errors-Refinement.md

```diff
@@ -0,0 +1,981 @@
+  class A ... { var c: T }
+}
+```
+
+Following the general rule that declarations in the base module are replaced by more specific declarations of the same kind in the refining module,
+a `var` declaration in a refining class must replace a `var` declaration in the class in the base (with the same type).
+
+
+## **Error: a field declaration (_name_) in a refining class (_class_) must repeat the syntactically same type as the field has in the refinement base** {#ref_mismatched_field_name}
+
+```dafny
+module P {
+  type T = int
+  class A { var c: int }
+}
+module Q refines P {
+  class A ... { var c: T }
+}
+```
+
+The field declarations in a refining class must have the same type as in the class in the base module. In fact, to enable easier checking that the type
+has not changed, the type must be expressed in the same syntactic form in the two declarations. It is not permitted to use a type in one place and 
+an equivalent type synonym in the other.
+
+
+## **Error: a field re-declaration (_name_) must be to ghostify the field** {#ref_refinement_field_must_add_ghost}
+
+```dafny
+module P {
+  class A { var c: int }
+}
+module Q refines P {
+  class A ... { var c: int }
+}
+```
+
+When a class is being refined, any field declaration in the base is copied into the refinement.
+If there is a redeclaration of the field, it must be to add a `ghost` modifier.
+
+## **Error: a _kind_ declaration (_name_) can only refine a _kind_** {#ref_mismatched_refinement_kind}
+
+```dafny
+module P {
+  function f(i: int): bool
+}
+module Q refines P {
+  predicate f(i: int) { true }
+}
+```
+
+The refining declaration must be the same kind of thing as the base declaration.
+For example both must be predicates or both must be functions (even if the function is one that returns a `bool`).
+
+## Error: a refining _kind_ is not allowed to add preconditions** {#ref_refinement_no_new_preconditions}
+
+```dafny
+module P {
+  predicate m(i: nat)
+}
+module Q refines P {
+  predicate m(i: nat) requires true { true }
+}
+```
+
+A function in a refining module must be able to be used in the same way as the abstract function in the base module.
+If there are additional preconditions, then the uses of the refined module may be more restricted.
+Thus no new preconditions may be added. This is a syntactic check, so no preconditions can be added even if they
+are implied by the existing preconditions.
+
+## **Error: a refining _kind_ is not allowed to extend the reads clause** {#ref_refinement_no_new_reads}
+
+```dafny
+module P {
+  predicate m() reads {}
+}
+module Q refines P {
+  predicate m() reads this {true }
+}
+```
+
+A function in a refining module must be able to be used in the same way as the abstract function in the base module.
+Extending the reads clause with additional objects cxhanges this equivalence and is not allowed.
+This change is syntactic: the refining function is not allowed to write any reads clauses; it just inherits those from
```

Comment:
```suggestion
This change is syntactic: the refining function is not allowed to write any reads clauses, it just inherits those from
```

Reply:
Actually the ; is correct and the comma not. But I changed it to avoid a series of ;-separated clauses.